# Charlie Rousmaniere - Unit 2 Project

## Main Question:
How do stem majors compare to non-stem majors?


## Import data and libraries
Items Covered:
* read_csv()
* describe()
* head()
* set index_col


In [2]:
import pandas as pd

grads = pd.read_csv('recent-grads.csv', index_col='Rank')

# Describing Data

In [3]:
grads.describe()

Major_code          Total            Men          Women  ShareWomen  \
count   173.000000     172.000000     172.000000     172.000000  172.000000   
mean   3879.815029   39370.081395   16723.406977   22646.674419    0.522223   
std    1687.753140   63483.491009   28122.433474   41057.330740    0.231205   
min    1100.000000     124.000000     119.000000       0.000000    0.000000   
25%    2403.000000    4549.750000    2177.500000    1778.250000    0.336026   
50%    3608.000000   15104.000000    5434.000000    8386.500000    0.534024   
75%    5503.000000   38909.750000   14631.000000   22553.750000    0.703299   
max    6403.000000  393735.000000  173809.000000  307087.000000    0.968954   

       Sample_size       Employed      Full_time      Part_time  \
count   173.000000     173.000000     173.000000     173.000000   
mean    356.080925   31192.763006   26029.306358    8832.398844   
std     618.361022   50675.002241   42869.655092   14648.179473   
min       2.000000       0.000000     111.000000       0.000000   
25%      39.000000    3608.000000    3154.000000    1030.000000   
50%     130.000000   11797.000000   10048.000000    3299.000000   
75%     338.000000   31433.000000   25147.000000    9948.000000   
max    4212.000000  307933.000000  251540.000000  115172.000000   

       Full_time_year_round    Unemployed  Unemployment_rate         Median  \
count            173.000000    173.000000         173.000000     173.000000   
mean           19694.427746   2416.329480           0.068191   40151.445087   
std            33160.941514   4112.803148           0.030331   11470.181802   
min              111.000000      0.000000           0.000000   22000.000000   
25%             2453.000000    304.000000           0.050306   33000.000000   
50%             7413.000000    893.000000           0.067961   36000.000000   
75%            16891.000000   2393.000000           0.087557   45000.000000   
max           199897.000000  28169.000000           0.177226  110000.000000   

              P25th          P75th   College_jobs  Non_college_jobs  \
count    173.000000     173.000000     173.000000        173.000000   
mean   29501.445087   51494.219653   12322.635838      13284.497110   
std     9166.005235   14906.279740   21299.868863      23789.655363   
min    18500.000000   22000.000000       0.000000          0.000000   
25%    24000.000000   42000.000000    1675.000000       1591.000000   
50%    27000.000000   47000.000000    4390.000000       4595.000000   
75%    33000.000000   60000.000000   14444.000000      11783.000000   
max    95000.000000  125000.000000  151643.000000     148395.000000   

       Low_wage_jobs  
count     173.000000  
mean     3859.017341  
std      6944.998579  
min         0.000000  
25%       340.000000  
50%      1231.000000  
75%      3466.000000  
max     48207.000000

In [4]:
grads.head()

Major_code                                      Major    Total      Men  \
Rank                                                                            
1           2419                      PETROLEUM ENGINEERING   2339.0   2057.0   
2           2416             MINING AND MINERAL ENGINEERING    756.0    679.0   
3           2415                  METALLURGICAL ENGINEERING    856.0    725.0   
4           2417  NAVAL ARCHITECTURE AND MARINE ENGINEERING   1258.0   1123.0   
5           2405                       CHEMICAL ENGINEERING  32260.0  21239.0   

        Women Major_category  ShareWomen  Sample_size  Employed  Full_time  \
Rank                                                                         
1       282.0    Engineering    0.120564           36      1976       1849   
2        77.0    Engineering    0.101852            7       640        556   
3       131.0    Engineering    0.153037            3       648        558   
4       135.0    Engineering    0.107313           16       758       1069   
5     11021.0    Engineering    0.341631          289     25694      23170   

      Part_time  Full_time_year_round  Unemployed  Unemployment_rate  Median  \
Rank                                                                           
1           270                  1207          37           0.018381  110000   
2           170                   388          85           0.117241   75000   
3           133                   340          16           0.024096   73000   
4           150                   692          40           0.050125   70000   
5          5180                 16697        1672           0.061098   65000   

      P25th   P75th  College_jobs  Non_college_jobs  Low_wage_jobs  
Rank                                                                
1     95000  125000          1534               364            193  
2     55000   90000           350               257             50  
3     50000  105000           456               176              0  
4     43000   80000           529               102              0  
5     50000   75000         18314              4440            972

In [5]:
grads.shape

(173, 20)

In [57]:
grads.min(axis=0)

Major_code                                         1100
Major                                        ACCOUNTING
Total                                             124.0
Men                                               119.0
Women                                               0.0
Major_category          Agriculture & Natural Resources
ShareWomen                                          0.0
Sample_size                                           2
Employed                                              0
Full_time                                           111
Part_time                                             0
Full_time_year_round                                111
Unemployed                                            0
Unemployment_rate                                   0.0
Median                                            22000
P25th                                             18500
P75th                                             22000
College_jobs                                    

In [56]:
grads.max(axis=0)

Major_code                        6403
Major                          ZOOLOGY
Total                         393735.0
Men                           173809.0
Women                         307087.0
Major_category          Social Science
ShareWomen                    0.968954
Sample_size                       4212
Employed                        307933
Full_time                       251540
Part_time                       115172
Full_time_year_round            199897
Unemployed                       28169
Unemployment_rate             0.177226
Median                          110000
P25th                            95000
P75th                           125000
College_jobs                    151643
Non_college_jobs                148395
Low_wage_jobs                    48207
dtype: object

In this section I ingested data from the csv file into a Pandas dataframe. I then used describe and head to get a general overview of the data. Shape also showed that there are 173 observations and 20 variables.

# Cleaning data

Items Covered:
* isnull()
* .any
* fillna()
* loc[]

In [6]:
# Vizualize null data
null_mask_rows = grads.isnull().any(axis=1)

null_rows = grads[null_mask_rows]

null_rows.head()

Major_code         Major  Total  Men  Women  \
Rank                                                
22          1104  FOOD SCIENCE    NaN  NaN    NaN   

                       Major_category  ShareWomen  Sample_size  Employed  \
Rank                                                                       
22    Agriculture & Natural Resources         NaN           36      3149   

      Full_time  Part_time  Full_time_year_round  Unemployed  \
Rank                                                           
22         2558       1121                  1735         338   

      Unemployment_rate  Median  P25th  P75th  College_jobs  Non_college_jobs  \
Rank                                                                            
22             0.096931   53000  32000  70000          1183              1274   

      Low_wage_jobs  
Rank                 
22              485

isnull created a boolean flag for all rows that had null values. By passing this boolean flag into the dataframe I found that only the Food Science major had missing values. 

In [7]:
# Replace null data

categorical_cols = ['Major', 'Major_category']

grads_no_missing = grads.fillna(grads.median(skipna=True, numeric_only=True))

grads_no_missing.loc[grads_no_missing['Major'] == 'FOOD SCIENCE']

Major_code         Major    Total     Men   Women  \
Rank                                                      
22          1104  FOOD SCIENCE  15104.0  5434.0  8386.5   

                       Major_category  ShareWomen  Sample_size  Employed  \
Rank                                                                       
22    Agriculture & Natural Resources    0.534024           36      3149   

      Full_time  Part_time  Full_time_year_round  Unemployed  \
Rank                                                           
22         2558       1121                  1735         338   

      Unemployment_rate  Median  P25th  P75th  College_jobs  Non_college_jobs  \
Rank                                                                            
22             0.096931   53000  32000  70000          1183              1274   

      Low_wage_jobs  
Rank                 
22              485

Using fillna, the missing values in food science were replaced with the median value from that column.

# Categorizing data

Items Coverd:
* value_counts()
* isin()
* adding columns
* groupby()

In [9]:
grads_no_missing['Major_category'].unique()

array(['Engineering', 'Business', 'Physical Sciences',
       'Law & Public Policy', 'Computers & Mathematics',
       'Agriculture & Natural Resources',
       'Industrial Arts & Consumer Services', 'Arts', 'Health',
       'Social Science', 'Biology & Life Science', 'Education',
       'Humanities & Liberal Arts', 'Psychology & Social Work',
       'Communications & Journalism', 'Interdisciplinary'], dtype=object)

In [10]:
Stem_majors = ['Engineering','Physical Sciences', 'Health', 
               'Computer & Mathematics', 'Biology & Life Science', 
               'Agriculture & Natural Resources']

Non_stem_majors = ['Buisiness', 'Law & Public Policy', 'Arts', 
                   'Social Science', 'Education', 'Humanities & Liberal Arts',
                      'Psychology & Social Work', 'Industrial Arts & Consumer Services',
                         'Communications & Journalism', 'Interdisciplinary']

grads_no_missing['STEM'] = ['STEM' if i in Stem_majors else 'Non-STEM' for i in grads_no_missing['Major_category'] ]

grads_no_missing['STEM'].value_counts()

STEM
Non-STEM    98
STEM        75
Name: count, dtype: int64

I created a new column in the dataframe called STEM which defined whether or not the major was a STEM major. From Value_Counts I can see that there are 98 non-stem majors and 75 stem majors.

In [17]:
grads_isSTEM = grads_no_missing.groupby('STEM')

grads_isSTEM.size()

STEM
Non-STEM    98
STEM        75
dtype: int64

This is another way of doing the value_counts function using groupby.

# Salary Analysis
Items Coverd:
* groupby()
* .agg()
* .apply()
* lambda
* loc()
* idxmax()
* mean, median, min, max,, var, std


In [13]:
# Max median salary by major in both stem and non-stem majors

highest_salary = grads_no_missing.groupby(['STEM']).apply(lambda df: df.loc[df.Median.idxmax()])

highest_salary

/tmp/ipykernel_1987/2956430514.py:3: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  highest_salary = grads_no_missing.groupby(['STEM']).apply(lambda df: df.loc[df.Median.idxmax()])


Major_code                  Major   Total     Men   Women  \
STEM                                                                  
Non-STEM        6202      ACTUARIAL SCIENCE  3777.0  2110.0  1667.0   
STEM            2419  PETROLEUM ENGINEERING  2339.0  2057.0   282.0   

         Major_category  ShareWomen  Sample_size  Employed  Full_time  ...  \
STEM                                                                   ...   
Non-STEM       Business    0.441356           51      2912       2924  ...   
STEM        Engineering    0.120564           36      1976       1849  ...   

          Full_time_year_round  Unemployed  Unemployment_rate  Median  P25th  \
STEM                                                                           
Non-STEM                  2482         308           0.095652   62000  53000   
STEM                      1207          37           0.018381  110000  95000   

           P75th  College_jobs  Non_college_jobs  Low_wage_jobs      STEM  
STEM                                                                       
Non-STEM   72000          1768               314            259  Non-STEM  
STEM      125000          1534               364            193      STEM  

[2 rows x 21 columns]

This cell shows that the highest median salary in non-stem majors is $62,000 (Actuarial Science) and the highest median salary in stem majors is $110,000 (Petroleum Engineering)

In [32]:
salaries = grads_no_missing.groupby('STEM').Median.agg(Average='mean', STD='std', Minimum = 'min', Med = 'median', Maximum='max', Variance='var')

salaries

Average           STD  Minimum      Med  Maximum      Variance
STEM                                                                         
Non-STEM  36146.938776   7254.834209    22000  35000.0    62000  5.263262e+07
STEM      45384.000000  13704.813122    26000  40000.0   110000  1.878219e+08

This cell creates columns which calculate the mean, standard deviation, minimum, median, variance, and maximum for both STEM and non-STEM majors. The minimum and median slaries are fairly similar for both stem and non-stem majors however the maximum median salaries for stem majors is much higher than the maximum salries for non-stem majors. The stem-majors also have more variation in their salaries.

In [47]:
grads_no_missing['Income_Range'] = pd.qcut(grads_no_missing['Median'], 3, labels=['low', 'med','high'])

grads_no_missing.groupby('STEM').Income_Range.value_counts()

STEM      Income_Range
Non-STEM  low             48
          med             29
          high            21
STEM      high            36
          med             27
          low             12
Name: count, dtype: int64


This cell uses qcut, which divides a column up into even bins and allows you to label these bins. In this case I divided the median income column into 3 bins and labled them low, med, and high. I then sorted this values by stem and non-stem majors and looked at the value counts. Stem majors had the vast majority of high incomes and a much lower proportion of low incomes. Non-stem majors were the opposite.

Its clear to see from this analysis that typically STEM majors have better median salaries than non-stem majors.

# Gender Analysis


In [49]:
grads_no_missing = grads_no_missing.assign(Percentage_Men= lambda df: df.Men / df.Total)

grads_no_missing.groupby('STEM').Percentage_Men.agg('mean')


STEM
Non-STEM    0.424511
STEM        0.545803
Name: Percentage_Men, dtype: float64

This cell assigned a new column to calculate the percentage of men in each major. I then used groupby to find the average percentage of men in stem majors and non-stem majors. There is a clear gender gap between stem and non-stem majors. Stem majors have 54% men whereas non-stem majors have only 42% men. 

In [52]:
lowest_men = grads_no_missing.groupby(['STEM']).apply(lambda df: df.loc[df.Percentage_Men.idxmin()])

lowest_men

/tmp/ipykernel_1987/1329962454.py:1: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  lowest_men = grads_no_missing.groupby(['STEM']).apply(lambda df: df.loc[df.Percentage_Men.idxmin()])


Major_code                                          Major    Total  \
STEM                                                                           
Non-STEM        2307                      EARLY CHILDHOOD EDUCATION  37589.0   
STEM            6102  COMMUNICATION DISORDERS SCIENCES AND SERVICES  38279.0   

             Men    Women Major_category  ShareWomen  Sample_size  Employed  \
STEM                                                                          
Non-STEM  1167.0  36422.0      Education    0.968954          342     32551   
STEM      1225.0  37054.0         Health    0.967998           95     29763   

          Full_time  ...  Unemployment_rate  Median  P25th  P75th  \
STEM                 ...                                            
Non-STEM      27569  ...           0.040105   28000  21000  35000   
STEM          19975  ...           0.047584   28000  20000  40000   

          College_jobs  Non_college_jobs  Low_wage_jobs      STEM  \
STEM                                                                
Non-STEM         23515              7705           2868  Non-STEM   
STEM             19957              9404           5125      STEM   

          Percentage_Men  Income_Range  
STEM                                    
Non-STEM        0.031046           low  
STEM            0.032002           low  

[2 rows x 23 columns]

The most female dominated majors in STEM and non-stem are Communication Disorders Sciences and Services, and Early Childhood Education respectivly.

In [53]:
highest_men = grads_no_missing.groupby(['STEM']).apply(lambda df: df.loc[df.Percentage_Men.idxmax()])

highest_men

/tmp/ipykernel_1987/1170889109.py:1: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  highest_men = grads_no_missing.groupby(['STEM']).apply(lambda df: df.loc[df.Percentage_Men.idxmax()])


Major_code                                        Major   Total  \
STEM                                                                        
Non-STEM        3801                        MILITARY TECHNOLOGIES   124.0   
STEM            2504  MECHANICAL ENGINEERING RELATED TECHNOLOGIES  4790.0   

             Men  Women                       Major_category  ShareWomen  \
STEM                                                                       
Non-STEM   124.0    0.0  Industrial Arts & Consumer Services    0.000000   
STEM      4419.0  371.0                          Engineering    0.077453   

          Sample_size  Employed  Full_time  ...  Unemployment_rate  Median  \
STEM                                        ...                              
Non-STEM            4         0        111  ...           0.000000   40000   
STEM               71      4186       4175  ...           0.056357   40000   

          P25th  P75th  College_jobs  Non_college_jobs  Low_wage_jobs  \
STEM                                                                    
Non-STEM  40000  40000             0                 0              0   
STEM      27000  52000          1861              2121            406   

              STEM  Percentage_Men  Income_Range  
STEM                                              
Non-STEM  Non-STEM        1.000000           med  
STEM          STEM        0.922547           med  

[2 rows x 23 columns]

The most Male dominated fields are Military technologies and Mechanical Engineering Related Technologies.